tree/dataframe/src/RDFHistoModels.cxx

```diff
@@ -213,6 +218,71 @@ TH3DModel::~TH3DModel()
 {
 }
 
+THnDModel::THnDModel(const ::THnD &h)
+   : fName(h.GetName()), fTitle(h.GetTitle()), fDim(h.GetNdimensions()), fNbins(fDim), fXmin(fDim), fXmax(fDim),
+     fBinEdges(fDim)
+{
+   for (int idim = 0; idim < fDim; ++idim) {
+      fNbins[idim] = h.GetAxis(idim)->GetNbins();
+      SetAxisProperties(h.GetAxis(idim), fXmin[idim], fXmax[idim], fBinEdges[idim]);
+   }
+}
+
+THnDModel::THnDModel(const char *name, const char *title, int dim, const int *nbins, const double *xmin,
+                     const double *xmax)
+   : fName(name), fTitle(title), fDim(dim), fBinEdges(dim)
+{
+   fNbins.reserve(fDim);
+   fXmin.reserve(fDim);
+   fXmax.reserve(fDim);
+   for (int idim = 0; idim < fDim; ++idim) {
+      fNbins.push_back(nbins[idim]);
+      fXmin.push_back(xmin[idim]);
+      fXmax.push_back(xmax[idim]);
+   }
+}
+
+THnDModel::THnDModel(const char *name, const char *title, int dim, const std::vector<int> &nbins,
+                     const std::vector<double> &xmin, const std::vector<double> &xmax)
+   : fName(name), fTitle(title), fDim(dim), fNbins(nbins), fXmin(xmin), fXmax(xmax), fBinEdges(dim)
+{
+}
+
+THnDModel::THnDModel(const char *name, const char *title, int dim, const int *nbins,
+                     const std::vector<std::vector<double>> &xbins)
+   : fName(name), fTitle(title), fDim(dim), fXmin(dim, 0.), fXmax(dim, 64.), fBinEdges(xbins)
+{
+   fNbins.reserve(fDim);
+   for (int idim = 0; idim < fDim; ++idim) {
+      fNbins.push_back(nbins[idim]);
+   }
+}
+
+THnDModel::THnDModel(const char *name, const char *title, int dim, const std::vector<int> &nbins,
+                     const std::vector<std::vector<double>> &xbins)
+   : fName(name), fTitle(title), fDim(dim), fNbins(nbins), fXmin(dim, 0.), fXmax(dim, 64.), fBinEdges(xbins)
+{
+}
+
+std::shared_ptr<::THnD> THnDModel::GetHistogram() const
+{
+   bool varbinning = false;
+   for (const auto &bins : fBinEdges) {
+      if (bins.size()) {
+         varbinning = true;
+         break;
```

Comment:
Is it really allowed to have some dimensions with zero bins in thns with variable binning?

Reply:
Hmm, I THINK this might actually be valid? (an axis with a single bin edge, and only an underflow and overflow bin)

In any case the logic here is the same as for TH2/3 (all of the arrays empty -> regular binning, otherwise variable binning)

Reply:
alright, consistency wins :+1: